core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -542,14 +556,14 @@ assert equalType(
                     equivalents.remove(slots[i], equi.iterator().next());
                   }
                 }
-                assert result.result.rowType.equals(result.call.query.rowType)
-                    : Pair.of(result.result, result.call.query);
-                equivalents.put(result.result, result.call.query);
+                equivalents.put(result1, result.call.query);
                 if (targetDescendant == target) {
                   // A real substitution happens. We purge the attempted
                   // replacement list and add them into substitution list.
                   // Meanwhile we stop matching the descendants and jump
                   // to the next subtree in pre-order traversal.
+                  assert result1.rowType.equals(result.call.query.rowType)
+                      : Pair.of(result1, result.call.query);
```

Comment:
This assert is not that necessary, because it is always true now.

Reply:
OK， I will remove it.